src/algebra/opposites.lean

```diff
@@ -69,65 +69,59 @@ variable {α}
 @[simp] lemma op_inv [has_inv α] (x : α) : op (x⁻¹) = (op x)⁻¹ := rfl
 @[simp] lemma unop_inv [has_inv α] (x : αᵒᵖ) : unop (x⁻¹) = (unop x)⁻¹ := rfl
 
-@[simp] lemma op_sub [add_group α] (x y : α) : op (x - y) = op x - op y := rfl
-@[simp] lemma unop_sub [add_group α] (x y : αᵒᵖ) : unop (x - y) = unop x - unop y := rfl
+@[simp] lemma op_sub [has_sub α] (x y : α) : op (x - y) = op x - op y := rfl
+@[simp] lemma unop_sub [has_sub α] (x y : αᵒᵖ) : unop (x - y) = unop x - unop y := rfl
 
 @[simp] lemma op_smul {R : Type*} [has_scalar R α] (c : R) (a : α) : op (c • a) = c • op a := rfl
 @[simp] lemma unop_smul {R : Type*} [has_scalar R α] (c : R) (a : αᵒᵖ) :
   unop (c • a) = c • unop a := rfl
 
 end
 
-instance [add_semigroup α] : add_semigroup (opposite α) :=
-{ add_assoc := λ x y z, unop_injective $ add_assoc (unop x) (unop y) (unop z),
-  .. opposite.has_add α }
+instance [add_semigroup α] : add_semigroup (αᵒᵖ) :=
+unop_injective.add_semigroup _ (λ x y, rfl)
 
 instance [add_left_cancel_semigroup α] : add_left_cancel_semigroup (opposite α) :=
-{ add_left_cancel := λ x y z H, unop_injective $ add_left_cancel $ op_injective H,
-  .. opposite.add_semigroup α }
+unop_injective.add_left_cancel_semigroup _ (λ x y, rfl)
 
 instance [add_right_cancel_semigroup α] : add_right_cancel_semigroup (opposite α) :=
-{ add_right_cancel := λ x y z H, unop_injective $ add_right_cancel $ op_injective H,
-  .. opposite.add_semigroup α }
+unop_injective.add_right_cancel_semigroup _ (λ x y, rfl)
 
 instance [add_comm_semigroup α] : add_comm_semigroup (opposite α) :=
 { add_comm := λ x y, unop_injective $ add_comm (unop x) (unop y),
   .. opposite.add_semigroup α }
 
-
 instance [nontrivial α] : nontrivial (opposite α) :=
-let ⟨x, y, h⟩ := exists_pair_ne α in nontrivial_of_ne (op x) (op y) (op_injective.ne h)
+op_injective.nontrivial
 
-section
-local attribute [semireducible] opposite
 @[simp] lemma unop_eq_zero_iff {α} [has_zero α] (a : αᵒᵖ) : a.unop = (0 : α) ↔ a = (0 : αᵒᵖ) :=
-iff.refl _
+unop_injective.eq_iff' rfl
```

Comment:
Doesn't this change this lemma's eligibility for `dsimp`?

Reply:
AFAIK, `dsimp` doesn't use `iff.refl _` lemmas, only `rfl` ones. Also, I'll probably change the definition of `opposite` to a structure, so this will not be an `iff.rfl` anymore.

Reply:
The downside of making `opposite` a structure is that `op (unop a) = a` stops being true by `rfl`. Either way, that's a discussion for another PR, but I think it's likely not so good an idea, for the same reason that changing `multiplicative` and `additive` to work like that wasn't.